Submodule Loop contains modified content
diff --git a/Loop/WatchApp/Base.lproj/Interface.storyboard b/Loop/WatchApp/Base.lproj/Interface.storyboard
index 03cdd19a..df4d3e1c 100644
--- a/Loop/WatchApp/Base.lproj/Interface.storyboard
+++ b/Loop/WatchApp/Base.lproj/Interface.storyboard
@@ -18,7 +18,7 @@
         <!--Loop-->
         <scene sceneID="gQv-2Z-iCt">
             <objects>
-                <controller identifier="ActionHUDController" title="Loop" contentMode="scaleAspectFit" spacing="7" id="rNf-Mh-tID" customClass="ActionHUDController" customModule="WatchApp_Extension">
+                <controller identifier="ActionHUDController" contentMode="scaleAspectFit" spacing="7" id="rNf-Mh-tID" customClass="ActionHUDController" customModule="WatchApp_Extension">
                     <items>
                         <group width="1" alignment="left" id="1wc-65-vel">
                             <items>
