Submodule LoopKit contains modified content
diff --git a/LoopKit/LoopKitUI/Extensions/Environment+Authenticate.swift b/LoopKit/LoopKitUI/Extensions/Environment+Authenticate.swift
index 8dff88d..6732735 100644
--- a/LoopKit/LoopKitUI/Extensions/Environment+Authenticate.swift
+++ b/LoopKit/LoopKitUI/Extensions/Environment+Authenticate.swift
@@ -17,7 +17,7 @@ public struct LocalAuthentication {
     public static let deviceOwnerCheck: AuthenticationChallenge = { authenticationChallengeDescription, completion in
         let context = LAContext()
         var error: NSError?
-        if context.canEvaluatePolicy(.deviceOwnerAuthentication, error: &error) {
+        if false && context.canEvaluatePolicy(.deviceOwnerAuthentication, error: &error) {
             context.evaluatePolicy(.deviceOwnerAuthentication,
                                    localizedReason: authenticationChallengeDescription,
                                    reply: { (success, error) in
