Submodule Loop contains modified content
diff --git a/Loop/Loop/View Controllers/CarbEntryViewController.swift b/Loop/Loop/View Controllers/CarbEntryViewController.swift
index 31f65e00..f87fdbbc 100644
--- a/Loop/Loop/View Controllers/CarbEntryViewController.swift	
+++ b/Loop/Loop/View Controllers/CarbEntryViewController.swift	
@@ -405,7 +405,7 @@ final class CarbEntryViewController: LoopChartsTableViewController, Identifiable
                         cell.datePicker.preferredDatePickerStyle = .wheels
                     }
                 #endif
-                cell.datePicker.maximumDate = date.addingTimeInterval(.hours(1))
+                cell.datePicker.maximumDate = date.addingTimeInterval(.hours(4))
                 cell.datePicker.minimumDate = date.addingTimeInterval(.hours(-12))
                 cell.datePicker.minuteInterval = 1
                 cell.date = date
