Submodule Loop contains modified content
diff --git a/Loop/Loop/Managers/LoopDataManager.swift b/Loop/Loop/Managers/LoopDataManager.swift
index 18a08166..0571c5bb 100644
--- a/Loop/Loop/Managers/LoopDataManager.swift
+++ b/Loop/Loop/Managers/LoopDataManager.swift
@@ -837,7 +837,11 @@ extension LoopDataManager {
 
             // Partial application factor assumes 5 minute intervals. If our looping intervals are shorter, then this will be adjusted
             self.timeBasedDoseApplicationFactor = 1.0
-            if let lastLoopCompleted = self.lastLoopCompleted {
+            // Limit Loop frequency to 4.8 minutes
+            if let lastLoopCompleted = self.lastLoopCompleted, Date().timeIntervalSince(lastLoopCompleted) < TimeInterval(minutes: 4.8) {
+                self.logger.default("Skipping loop attempt as last loop completed at %{public}@", String(describing: lastLoopCompleted))
+                return
+            } else if let lastLoopCompleted = self.lastLoopCompleted {
                 let timeSinceLastLoop = max(0, Date().timeIntervalSince(lastLoopCompleted))
                 self.timeBasedDoseApplicationFactor = min(1, timeSinceLastLoop/TimeInterval.minutes(5))
                 self.logger.default("Looping with timeBasedDoseApplicationFactor = %{public}@", String(describing: self.timeBasedDoseApplicationFactor))
